compiler/src/dmd/func.d

```diff
@@ -1510,6 +1510,12 @@ extern (C++) class FuncDeclaration : Declaration
      */
     extern (D) final bool setImpure(Loc loc = Loc.init, const(char)* fmt = null, RootObject arg0 = null)
     {
+        if (purityInprocess && semanticRun < PASS.semantic3 && _scope)
```

Comment:
Is this the right place to run the semantic analysis? If a function needs the sema to have run it seems messy to do it inside the function.

Reply:
It shouldn't be, but this was the similar fix for `@nogc` inference. Would be great to have a concise fix for this tho :/